File: hub/src/test/unit/java/com/thoughtworks/selenium/grid/hub/ServletParameterAdapterTest.java
package com.thoughtworks.selenium.grid.hub;

import com.thoughtworks.selenium.grid.HttpParameters;
import static junit.framework.Assert.assertEquals;
import static junit.framework.Assert.assertTrue;
import org.junit.Test;

import java.util.Collections;
import java.util.HashMap;
import java.util.Map;


public class ServletParameterAdapterTest {

    @Test
    public void parametersAreCopiedWhenAMapIsProvidedToConstructor() {
        final Map<String, String[]> servletParameterMap;

        servletParameterMap = new HashMap<String, String[]>();
        servletParameterMap.put("a name", new String[] { "a value" });
        assertEquals("a value", new HttpParameters(servletParameterMap).get("a name"));
    }

    @Test
    public void getReturnsOnlyTheFirstValueOfAParameter() {
        final Map<String, String[]> servletParameterMap;

        servletParameterMap = new HashMap<String, String[]>();
        servletParameterMap.put("a name", new String[] { "first value", "second value" });
        assertEquals("first value", new HttpParameters(servletParameterMap).get("a name"));
    }

    @Test
    public void getReturnsNullWhenParameterValueIsAnEmptyArray() {
        final Map<String, String[]> servletParameterMap;

        servletParameterMap = new HashMap<String, String[]>();
        servletParameterMap.put("a name", new String[] { });
        assertEquals(null, new HttpParameters(servletParameterMap).get("a name"));
    }

    @Test
    public void putCanChangeAValueEventIfMapProvidedToConstructorIsFrozen() {
        final Map<String, String[]> servletParameterMap;
        final HttpParameters parameters;

        servletParameterMap = new HashMap<String, String[]>();
        servletParameterMap.put("a name", new String[] { "original value" });
        parameters = new HttpParameters(Collections.unmodifiableMap(servletParameterMap));
        parameters.put("a name", "new value");
        assertEquals("new value", parameters.get("a name"));
    }


    @Test
    public void namesIsEmptyWhenThereIsNoParameter() {
        assertTrue(new HttpParameters().names().isEmpty());
    }

    @Test
    public void namesContainsAllTheParameterNames() {
        final HttpParameters parameters = new HttpParameters();
        parameters.put("selenium", "grid");
        parameters.put("open", "qa");
        assertEquals(2, parameters.names().size());
        assertTrue(parameters.names().contains("selenium"));
        assertTrue(parameters.names().contains("open"));
    }

}

File: hub/src/test/unit/com/thoughtworks/selenium/grid/hub/servlet_parameter_adapter_test.rb
require File.expand_path(File.dirname(__FILE__) + "/../../../../../unit_test_helper")

include_class 'java.util.HashMap'
include_class 'com.thoughtworks.selenium.grid.hub.HttpParameters'

class ServletParameterAdapterTest < Test::Unit::TestCase

  test "toString just have class name when request has no parameter" do
    assert_equal "", HttpParameters.new.to_s
  end

  test "toString includes request parameters when any" do
    parameters = HttpParameters.new
    parameters.put "cmd", "testComplete"
    assert_equal 'cmd => "testComplete"', parameters.to_s
  end

  test "toString seperates parameters with space" do
    parameters = HttpParameters.new
    parameters.put "cmd", "testComplete"
    parameters.put "sessionId", "1234"
    assert_equal 'cmd => "testComplete", sessionId => "1234"', parameters.to_s
  end

  test "get returns null when parameter does not exist" do
      assert_equal nil, HttpParameters.new.get("unknown parameter")
  end

  test "get returns nil when parameter value is null" do
    parameters = HttpParameters.new
    parameters.put "a name", nil
    assert_equal nil, parameters.get("a name")
  end

end